{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "9d8be909",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import re\n",
    "\n",
    "raw = pd.read_csv('./bioinf-edit1 - All.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "2236d31b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Mating pathway\n",
    "mating_pathway = {\n",
    "    \"Ste2\",\n",
    "    \"Ste3\",\n",
    "    \"Gαβγ\",\n",
    "    \"Ste5\",\n",
    "    \"Ste4\",\"Ste1\",\"Gpa1\",\n",
    "    \"Cdc24\",\n",
    "    \"Cdc42\",\n",
    "    \"Ste20\",\n",
    "    \"Ste50\",\n",
    "    \"Ste11\",\n",
    "    \"Ste7\",\n",
    "    \"Fus3\",\n",
    "    \"Ste12\",\n",
    "}\n",
    "\n",
    "# Filamentous growth pathway\n",
    "filamentous_growth_pathway = {\n",
    "    \"Msb2\",\n",
    "    \"Sho1\",\"Opy2\",\n",
    "    \"Cdc24\",\n",
    "    \"Cdc42\",\n",
    "    \"Ste20\",\n",
    "    \"Ste50\",\n",
    "    \"Ste11\",\n",
    "    \"Ste7\",\n",
    "    \"Kss1\",\n",
    "    \"Ste12\",\"Tec1\",\n",
    "}\n",
    "\n",
    "# High osmolarity pathway\n",
    "high_osmolarity_pathway = {\n",
    "    \"Msb2\",\"Hkr1\",\n",
    "    \"Sho1\",\"Opy2\",\n",
    "    \"Cdc24\",\n",
    "    \"Cdc42\",\n",
    "    \"Ste20\",\n",
    "    \"Ste50\",\n",
    "    \"Ste11\",\n",
    "    \"Pbs2\",\n",
    "    \"Hog1\",\n",
    "    \"Hot1\",\n",
    "}\n",
    "\n",
    "# Combine into a dictionary for easy access\n",
    "pathways_hm = {\n",
    "    \"Mating\": mating_pathway,\n",
    "    \"Filamentous Growth\": filamentous_growth_pathway,\n",
    "    \"High Osmolarity\": high_osmolarity_pathway\n",
    "}\n",
    "pathways = set([x.lower() for x in mating_pathway.union(high_osmolarity_pathway).union(filamentous_growth_pathway)])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "fdda15af",
   "metadata": {},
   "outputs": [],
   "source": [
    "def extract_gene_name(info_field):\n",
    "    if \"ANN=\" in info_field:\n",
    "        ann_entries = info_field.split(\"ANN=\")[1].split(\";\")[0]\n",
    "        annotations = ann_entries.split(\",\")\n",
    "        gene_info = set()\n",
    "        for ann in annotations:\n",
    "            ann_fields = ann.split(\"|\")\n",
    "            if len(ann_fields) > 4:\n",
    "                gene_name = ann_fields[3]\n",
    "                #gene_id = ann_fields[4]\n",
    "                gene_info.add(gene_name.lower())\n",
    "        return gene_info\n",
    "    else:\n",
    "        return set() \n",
    "\n",
    "def augument_data(data):\n",
    "  data = data.replace('./.:.',  '')\n",
    "  sorting_table = {\"HIGH\":0, \"MODERATE\":1, \"LOW\":2, 'MODIFIER':3}\n",
    "  data['Severity'] = data['FORMAT'].apply(lambda x: sorting_table[x.split('|')[2]])\n",
    "  data['GENES_AFFECTED'] = data['FORMAT'].map(extract_gene_name)\n",
    "  #data['Gene'] = data['FORMAT'].apply(lambda x: x.split('|')[3])\n",
    "  return data\n",
    "\n",
    "data = augument_data(raw)\n",
    "\n",
    "strains = list(map(lambda x : x+'.bam', ['G28c1', 'D11c1', 'I38c2', 'M59c1', 'WT']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "750f7161",
   "metadata": {},
   "outputs": [],
   "source": [
    "def analysis(strain, quality_threshold = 100 ):\n",
    "  is_homozygous = lambda x: x.startswith('1/1')\n",
    "  others = strains.copy() # strains = list(map(lambda x : x+'.bam', ['G28c1', 'D11c1', 'I38c2', 'M59c1', 'WT']))\n",
    "  others.remove(strain)\n",
    "  promising_picks = data[\n",
    "    (data[strain].apply(is_homozygous)) &\n",
    "    (data[\"QUAL\"] > quality_threshold) &\n",
    "    (data[others].applymap(is_homozygous).sum(axis=1)<=2)].sort_values(by='Severity', ascending=True)\n",
    "  pcks = set()\n",
    "  for ga in promising_picks['GENES_AFFECTED']: pcks = pcks.union(ga)\n",
    "  cross = pathways.intersection(pcks)\n",
    "  if len(cross) > 0:\n",
    "    print(f'FOUND cross between pathway and affected gene in {strain}\\n\\t{cross}')\n",
    "  else:\n",
    "    print(f'FOUND nothing in {strain}')\n",
    "  return promising_picks\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "76fea559",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "FOUND nothing in G28c1.bam\n",
      "FOUND cross between pathway and affected gene in D11c1.bam\n",
      "\t{'ste12'}\n",
      "FOUND cross between pathway and affected gene in I38c2.bam\n",
      "\t{'kss1'}\n",
      "FOUND nothing in M59c1.bam\n",
      "FOUND nothing in WT.bam\n"
     ]
    }
   ],
   "source": [
    "for strain in strains:\n",
    "  analysis(strain)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "b3c53efe",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'cdc24',\n",
       " 'cdc42',\n",
       " 'fus3',\n",
       " 'gpa1',\n",
       " 'gαβγ',\n",
       " 'hkr1',\n",
       " 'hog1',\n",
       " 'hot1',\n",
       " 'kss1',\n",
       " 'msb2',\n",
       " 'opy2',\n",
       " 'pbs2',\n",
       " 'sho1',\n",
       " 'ste1',\n",
       " 'ste11',\n",
       " 'ste12',\n",
       " 'ste2',\n",
       " 'ste20',\n",
       " 'ste3',\n",
       " 'ste4',\n",
       " 'ste5',\n",
       " 'ste50',\n",
       " 'ste7',\n",
       " 'tec1'}"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pathways"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "a785965f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "M59c1.bam\n",
      "FOUND nothing in M59c1.bam\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>FILTER</th>\n",
       "      <th>INFO</th>\n",
       "      <th>FORMAT</th>\n",
       "      <th>Unnamed: 9</th>\n",
       "      <th>D11c1.bam</th>\n",
       "      <th>G28c1.bam</th>\n",
       "      <th>I38c2.bam</th>\n",
       "      <th>M59c1.bam</th>\n",
       "      <th>WT.bam</th>\n",
       "      <th>Severity</th>\n",
       "      <th>GENES_AFFECTED</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>370</th>\n",
       "      <td>XIV</td>\n",
       "      <td>668562</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>221.9990</td>\n",
       "      <td>.</td>\n",
       "      <td>VDB=0.43412;SGB=-0.69312;MQSB=1;MQ0F=0;AF1=1;A...</td>\n",
       "      <td>ANN=A|stop_gained|HIGH|YNR021W|YNR021W|transcr...</td>\n",
       "      <td>GT:PL</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1/1:255,96,0</td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "      <td>{ynr021w}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>389</th>\n",
       "      <td>XV</td>\n",
       "      <td>563576</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>221.9990</td>\n",
       "      <td>.</td>\n",
       "      <td>VDB=0.552091;SGB=-0.693141;MQSB=1;MQ0F=0;AF1=1...</td>\n",
       "      <td>ANN=T|stop_gained|HIGH|RGA1|YOR127W|transcript...</td>\n",
       "      <td>GT:PL</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1/1:255,111,0</td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "      <td>{rga1}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>437</th>\n",
       "      <td>Mito</td>\n",
       "      <td>65922</td>\n",
       "      <td>.</td>\n",
       "      <td>GGGGGC</td>\n",
       "      <td>G</td>\n",
       "      <td>214.4580</td>\n",
       "      <td>.</td>\n",
       "      <td>INDEL;IDV=1;IMF=0.00571429;VDB=1.50203e-08;SGB...</td>\n",
       "      <td>ANN=G|frameshift_variant|HIGH|Q0182|Q0182|tran...</td>\n",
       "      <td>GT:PL</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1/1:255,255,0</td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "      <td>{q0182}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>I</td>\n",
       "      <td>12169</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>77.0126</td>\n",
       "      <td>.</td>\n",
       "      <td>VDB=0.182083;SGB=-0.616816;RPB=0.931779;MQB=0....</td>\n",
       "      <td>ANN=T|missense_variant|MODERATE|YAL064W-B|YAL0...</td>\n",
       "      <td>GT:PL</td>\n",
       "      <td>0/1:87,0,4</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1/1:107,3,0</td>\n",
       "      <td></td>\n",
       "      <td>1</td>\n",
       "      <td>{yal064w-b}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>299</th>\n",
       "      <td>XI</td>\n",
       "      <td>479905</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>221.9990</td>\n",
       "      <td>.</td>\n",
       "      <td>VDB=0.00732327;SGB=-0.691153;MQSB=1;MQ0F=0;AF1...</td>\n",
       "      <td>ANN=A|missense_variant|MODERATE|ALY1|YKR021W|t...</td>\n",
       "      <td>GT:PL</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1/1:255,54,0</td>\n",
       "      <td></td>\n",
       "      <td>1</td>\n",
       "      <td>{aly1}</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    #CHROM     POS ID     REF ALT      QUAL FILTER  \\\n",
       "370    XIV  668562  .       C   A  221.9990      .   \n",
       "389     XV  563576  .       G   T  221.9990      .   \n",
       "437   Mito   65922  .  GGGGGC   G  214.4580      .   \n",
       "19       I   12169  .       C   T   77.0126      .   \n",
       "299     XI  479905  .       T   A  221.9990      .   \n",
       "\n",
       "                                                  INFO  \\\n",
       "370  VDB=0.43412;SGB=-0.69312;MQSB=1;MQ0F=0;AF1=1;A...   \n",
       "389  VDB=0.552091;SGB=-0.693141;MQSB=1;MQ0F=0;AF1=1...   \n",
       "437  INDEL;IDV=1;IMF=0.00571429;VDB=1.50203e-08;SGB...   \n",
       "19   VDB=0.182083;SGB=-0.616816;RPB=0.931779;MQB=0....   \n",
       "299  VDB=0.00732327;SGB=-0.691153;MQSB=1;MQ0F=0;AF1...   \n",
       "\n",
       "                                                FORMAT Unnamed: 9   D11c1.bam  \\\n",
       "370  ANN=A|stop_gained|HIGH|YNR021W|YNR021W|transcr...      GT:PL               \n",
       "389  ANN=T|stop_gained|HIGH|RGA1|YOR127W|transcript...      GT:PL               \n",
       "437  ANN=G|frameshift_variant|HIGH|Q0182|Q0182|tran...      GT:PL               \n",
       "19   ANN=T|missense_variant|MODERATE|YAL064W-B|YAL0...      GT:PL  0/1:87,0,4   \n",
       "299  ANN=A|missense_variant|MODERATE|ALY1|YKR021W|t...      GT:PL               \n",
       "\n",
       "    G28c1.bam I38c2.bam      M59c1.bam WT.bam  Severity GENES_AFFECTED  \n",
       "370                       1/1:255,96,0                0      {ynr021w}  \n",
       "389                      1/1:255,111,0                0         {rga1}  \n",
       "437                      1/1:255,255,0                0        {q0182}  \n",
       "19                         1/1:107,3,0                1    {yal064w-b}  \n",
       "299                       1/1:255,54,0                1         {aly1}  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "s = 3\n",
    "print(strains[s])\n",
    "analysis(strains[s], quality_threshold=60)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
